src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,448 @@
+   }
+
+   /**
+    * Schedules {@code n} more executions.
+    * <p>
+    * If the task was not executing and {@code n > 0}, it will immediately begin executing.
+    * <p>
+    * You may also subtract from the number of scheduled executions by passing in a negative number.
+    * This method cannot cause the scheduled execution count to go below 0.
+    * <p>
+    * This method does not do anything if indefinitely repeating execution is scheduled.
+    *
+    * @param n If {@code n > 0}, the number of executions to schedule.
+    *          If {@code n < 0}, the number of executions to unschedule.
+    */
+   public void addScheduled(long n)
+   {
+      executionState.addScheduled(n);
+   }
+
+   /**
+    * @return The number of scheduled executions.
+    */
+   public long getScheduled()
+   {
+      return executionState.getScheduled();
+   }
+
+   /**
+    * @return Whether a task is currently executing.
+    */
+   public boolean isExecuting()
+   {
+      return executionState.isExecuting();
+   }
+
+   /**
+    * @return {@code true} if the thread has been started via {@link #start}
+    *          and not yet killed via {@link #kill} or {@link #blockingKill}.
+    */
+   public boolean isRunning()
+   {
+      return running;
+   }
+
+   /**
+    * @return The total number of task executions completed by this thread.
+    */
+   public long getCompleted()
+   {
+      return executionState.getCompleted();
+   }
+
+   /**
+    * Block until the next start of the task.
+    * @return {@code true} if interrupted.
+    */
+   public boolean blockUntilNextTaskExecution()
+   {
+      synchronized (executionState)
+      {
+         do
+         {
+            if (executionState.waitForChange())
+               return true;
+
+         } while (!executionState.isExecuting());
+      }
+
+      return false;
+   }
+
+   /**
+    * Block until the next completion of the task.
+    * @return {@code true} if interrupted.
+    */
+   public boolean blockUntilNextTaskCompletion()
+   {
+      synchronized (executionState)
+      {
+         long completedBefore = executionState.getCompleted();
+         while (completedBefore == executionState.getCompleted())
+         {
+            if (executionState.waitForChange())
+               return true;
+         }
+      }
+
+      return false;
+   }
+
+   /**
+    * Block until there are no scheduled task executions, which may be immediately.
+    * @return {@code true} if interrupted.
+    */
+   public boolean blockUntilNoScheduledTasks()
+   {
+      synchronized (executionState)
+      {
+         while (executionState.getScheduled() != 0 || executionState.isExecuting())
+         {
+            if (executionState.waitForChange())
+               return true;
+         }
+      }
+
+      return false;
+   }
+
+   /**
+    * Allows the thread to die.
+    * Any currently executing task will first be allowed to finish.
+    * This class cannot be reused after this point.
```

Comment:
Is there a way to ensure it can't be reused?

Reply:
Yes, that can work

Reply:
You'll already get one
![image](https://github.com/user-attachments/assets/f8097ab0-8670-4302-a95f-27452553688e)

Reply:
Might be worth adding a test to assert this

Reply:
Come to think of it, kill then start still works. Technically there's nothing wrong with that, except that it doesn't make sense